### PR TITLE
Add Druid e2e tests (I)

### LINKED ISSUE #194: [Feature] Standalone Helm Chart for Etcd Druid
**Feature (What you would like to be added):**
Would be great if we would have a standalone helm chart for druid which can be used outside Gardener.

**Motivation (Why is this needed?):**
Using druid instead of the etcd-operator and utilizing the backup and restore capabilities.

**Approach/Hint to the implement solution (optional):**
Cool thing would be to have a chart in the `charts` repository and released inside this github project via https://github.com/helm/chart-releaser-action in the repositories `gh-pages` branch.

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area testing
/kind enhancement

**What this PR does / why we need it**:
This PR adds e2e tests for Etcd-Druid. It uses the foundation of #195 with additional automation.

Please see `docs/development/local-e2e-tests.md` for more information about how to run those tests.

**Which issue(s) this PR fixes**:
Fixes #194
Fixes parts of #286

**Special notes for your reviewer**:
Co-authored-by: Shreyas Rao <shreyas.sriganesh.rao@sap.com>

ℹ️ Test machinery integration is not part of this PR and will follow as soon as the credential rotation feature is implemented in TM.

**Release note**:

```feature operator
A Helm chart for deploying Etcd-Druid is now available in `charts/druid`.
```
```feature developer
Developers can now run Druid e2e tests via `make test-e2e`. Please see `docs/development/local-e2e-tests.md` for detailed information.
```

### CODE REVIEW

**Comment by ishan16696** [comment]
Hi @timuthy ,
When e2e-tests fails they don't do the cleanup of resources like secrets, config-map etc .... Manually doing the cleanup of resources locally won't be an issue but it might be an issue if we gonna use this as TM test in ci pipeline ... right ?
Can we try to do the cleanup of resources irrespective of e2e tests fails or succeeds?

**Comment by ishan16696** [comment]
Hi @timuthy ,
I have retested this PR after you rebased on master.
The e2e tests fails with this error:
```
# github.com/gardener/etcd-druid/test/e2e [github.com/gardener/etcd-druid/test/e2e.test]
test/e2e/utils.go:127:3: cannot use "k8s.io/api/core/v1".SecretReference{...} (type "k8s.io/api/core/v1".SecretReference) as type "github.com/gardener/etcd-druid/api/v1alpha1".SecretReference in field value
test/e2e/utils.go:208:3: unknown field 'TLS' in struct literal of type "github.com/gardener/etcd-druid/api/v1alpha1".EtcdConfig
test/e2e/druid_test.go:59:19: etcd.Spec.Etcd.TLS undefined (type "github.com/gardener/etcd-druid/api/v1alpha1".EtcdConfig has no field or method TLS)
test/e2e/druid_test.go:60:19: etcd.Spec.Etcd.TLS undefined (type "github.com/gardener/etcd-druid/api/v1alpha1".EtcdConfig has no field or method TLS)
test/e2e/druid_test.go:61:19: etcd.Spec.Etcd.TLS undefined (type "github.com/gardener/etcd-druid/api/v1alpha1".EtcdConfig has no field or method TLS)
test/e2e/druid_test.go:698:19: etcd.Spec.Etcd.TLS undefined (type "github.com/gardener/etcd-druid/api/v1alpha1".EtcdConfig has no field or method TLS)
test/e2e/druid_test.go:699:19: etcd.Spec.Etcd.TLS undefined (type "github.com/gardener/etcd-druid/api/v1alpha1".EtcdConfig has no field or method TLS)
test/e2e/druid_test.go:700:19: etcd.Spec.Etcd.TLS undefined (type "github.com/gardener/etcd-druid/api/v1alpha1".EtcdConfig has no field or method TLS)
FAIL	github.com/gardener/etcd-druid/test/e2e [build failed]
FAIL
make: *** [test-e2e] Error 2
```

**Comment by timuthy** [comment]
Thanks for the feedback @ishan16696. I'll have a 👀

**Comment by timuthy** [comment]
The errors have been fixed with 350cc47efd8ff39fab63b97330eb3b0fb0215dd0. PTAL @ishan16696.

File: docs/development/local-e2e-tests.md [hunk 1]
```diff
@@ -33,12 +33,13 @@ The following providers are supported for e2e tests:
...
+| `setup`	  | Create a storage bucket which is used for etcd backups (only with cloud providers). 	                                                                                                                      |
+| `deploy`	 | Build Docker image, upload it to registry (if remote cluster - see [Docker build](https://skaffold.dev/docs/pipeline-stages/builders/docker/)), deploy Helm chart (`charts/druid`) to Kubernetes cluster.	 |
+| `test`    | Execute e2e tests as defined in `test/e2e`.                                                                                                                                                                |
+| `undeploy` | Remove the deployed artifacts from Kubernetes cluster.                                                                                                                                                     |
+| `cleanup` | Delete storage bucket and Druid deployment from test cluster.	                                                                                                                                             |
```
Comment (ishan16696) [inline docs/development/local-e2e-tests.md:42]: Isn't druid deployment get deleted in `undeploy` step ?

Comment (timuthy) [reply docs/development/local-e2e-tests.md:42]: Yes it is. The Druid `deployment` is one of the deployed artifacts 🤔

**Comment by ishan16696** [comment]
`GitGuardian Security Checks` is not having a green tick .... Is this expected ?

File: docs/development/local-e2e-tests.md (hunk 1 above)
Comment (ishan16696) [reply docs/development/local-e2e-tests.md:42]: ok, what I'm asking is `cleanup` also does that ? as it is mentioned `Delete storage bucket and Druid deployment from test cluster`

**Comment by ishan16696** [comment]
Is this https://github.com/gardener/etcd-druid/pull/296#issuecomment-1157262255 comment has been addressed or should we create a follow-up issue to track this ?

**Comment by timuthy** [comment]
> GitGuardian Security Checks is not having a green tick .... Is this expected ?

Yes, here it's because of the test TLS certificates and keys.

> Is this https://github.com/gardener/etcd-druid/pull/296#issuecomment-1157262255 has been addressed now or should we create a follow-up issue to track this just to delete the orphan resources if this has been enable it in a CI pipeline?

I think it's quite helpful that the cleanup does not happen when tests fail because it gives developers the opportunity to troubleshoot the system, e.g. checking the logs of etcd-druid. For the pipeline this shouldn't be a problem either because we can define a separate `make steps="cleanup" test-e2e` step that is executed in any case.

**Comment by ishan16696** [comment]
> For the pipeline this shouldn't be a problem either because we can define a separate make steps="cleanup" test-e2e step that is executed in any case.

ok, make sense

**Review by ishan16696 — APPROVED** [review]

**Comment by timuthy** [comment]
As discussed @ishan16696, some improvements were required through f1f6ae2c8296785f763d75b56fcd599438c5bd2e. PTAL.

**Review by ishan16696 — CHANGES_REQUESTED** [review]
Overrall, looks good to me. Couple of nitpicks.
And Can you also mention in `docs/development/local-e2e-tests.md` about cleaning up your existing resources of cluster-scope like CRD,clusterroles before running these e2e-tests.

**Comment by ishan16696** [comment]
Hey @timuthy ,
I have run the e2e-tests for Azure and AWS providers, so e2e-tests are running fine for both of the providers but one thing I noticed is `shoot` namespace is not getting deleted after cleanup step.
It also left some orphans resources in `shoot` namespace:
```
k -n shoot get secrets                                                                                                                 
NAME                  TYPE                DATA   AGE
ca-etcd-az            Opaque              2      2m36s
etcd-backup-az        Opaque              3      2m33s
etcd-client-tls-az    kubernetes.io/tls   3      2m36s
etcd-server-cert-az   kubernetes.io/tls   3      2m36s
```

**Comment by ishan16696** [comment]
yes, now it's getting cleanup, thanks for quick fix.
Would you like to address this as well https://github.com/gardener/etcd-druid/pull/296#pullrequestreview-1077141741 ?

File: test/e2e/etcd_backup_test.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+			provider := p
+			Context(fmt.Sprintf("with provider %s", provider.Name), func() {
+				BeforeEach(func() {
+					cl, err = getKubernetesClient(kubeconfigPath)
+					Expect(err).ShouldNot(HaveOccurred())
+
+					etcdName = fmt.Sprintf("etcd-%s", provider.Name)
+
+					if provider.Storage == nil || provider.Storage.Provider == providerLocal {
+						return
```
[the remark this answers is not in the document]
Comment (timuthy) [reply test/e2e/etcd_backup_test.go:65]: Sorry, this was a leftover and an unreachable piece of code. I'll remove it.

**Comment by timuthy** [comment]
Thanks for the feedback @ishan16696 🚀 I addressed all comments.

**Review by ishan16696 — APPROVED** [review]
LGTM!!

**Pull request merged**